File: infrastructure/go-sdk/circuit_breaker.go
package infra

import (
	"context"
	"errors"
	"fmt"
	"math"
	"math/rand"
	"net/http"
	"sync"
	"time"

	"go.uber.org/zap"
)

// Circuit breaker states.
const (
	StateClosed   = "closed"
	StateOpen     = "open"
	StateHalfOpen = "half-open"
)

var (
	ErrCircuitOpen    = errors.New("circuit breaker is open")
	ErrTooManyRetries = errors.New("max retries exceeded")
)

// CircuitBreaker implements the circuit breaker pattern for inter-service calls.
type CircuitBreaker struct {
	mu               sync.RWMutex
	name             string
	state            string
	failureCount     int
	successCount     int
	failureThreshold int
	successThreshold int
	timeout          time.Duration
	lastFailure      time.Time
	halfOpenMax      int
	halfOpenCurrent  int
	logger           *zap.Logger
}

// CircuitBreakerConfig configures a circuit breaker instance.
type CircuitBreakerConfig struct {
	Name             string
	FailureThreshold int
	SuccessThreshold int
	Timeout          time.Duration
	HalfOpenMax      int
}

// DefaultCircuitBreakerConfig returns production-ready defaults.
func DefaultCircuitBreakerConfig(name string) CircuitBreakerConfig {
	return CircuitBreakerConfig{
		Name:             name,
		FailureThreshold: 5,
		SuccessThreshold: 3,
		Timeout:          30 * time.Second,
		HalfOpenMax:      1,
	}
}

// NewCircuitBreaker creates a new circuit breaker.
func NewCircuitBreaker(logger *zap.Logger, cfg CircuitBreakerConfig) *CircuitBreaker {
	return &CircuitBreaker{
		name:             cfg.Name,
		state:            StateClosed,
		failureThreshold: cfg.FailureThreshold,
		successThreshold: cfg.SuccessThreshold,
		timeout:          cfg.Timeout,
		halfOpenMax:      cfg.HalfOpenMax,
		logger:           logger,
	}
}

// Execute runs the given function through the circuit breaker.
func (cb *CircuitBreaker) Execute(ctx context.Context, fn func(context.Context) error) error {
	if err := cb.canExecute(); err != nil {
		return err
	}

	err := fn(ctx)
	cb.recordResult(err)
	return err
}

func (cb *CircuitBreaker) canExecute() error {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case StateClosed:
		return nil
	case StateOpen:
		if time.Since(cb.lastFailure) > cb.timeout {
			cb.state = StateHalfOpen
			cb.halfOpenCurrent = 0
			cb.successCount = 0
			cb.logger.Info("circuit_breaker_half_open", zap.String("name", cb.name))
			return nil
		}
		return ErrCircuitOpen
	case StateHalfOpen:
		if cb.halfOpenCurrent >= cb.halfOpenMax {
			return ErrCircuitOpen
		}
		cb.halfOpenCurrent++
		return nil
	}
	return nil
}

func (cb *CircuitBreaker) recordResult(err error) {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if err != nil {
		cb.failureCount++
		cb.lastFailure = time.Now()

		if cb.state == StateHalfOpen {
			cb.state = StateOpen
			cb.logger.Warn("circuit_breaker_reopened", zap.String("name", cb.name))
			return
		}

		if cb.failureCount >= cb.failureThreshold {
			cb.state = StateOpen
			cb.logger.Warn("circuit_breaker_opened",
				zap.String("name", cb.name),
				zap.Int("failures", cb.failureCount),
			)
		}
		return
	}

	if cb.state == StateHalfOpen {
		cb.successCount++
		if cb.successCount >= cb.successThreshold {
			cb.state = StateClosed
			cb.failureCount = 0
			cb.successCount = 0
			cb.logger.Info("circuit_breaker_closed", zap.String("name", cb.name))
		}
		return
	}

	cb.failureCount = 0
}

// State returns the current circuit breaker state.
func (cb *CircuitBreaker) State() string {
	cb.mu.RLock()
	defer cb.mu.RUnlock()
	return cb.state
}

// RetryConfig configures retry behavior with exponential backoff.
type RetryConfig struct {
	MaxRetries  int
	BaseDelay   time.Duration
	MaxDelay    time.Duration
	Multiplier  float64
	JitterRatio float64
}

// DefaultRetryConfig returns production-ready retry defaults.
func DefaultRetryConfig() RetryConfig {
	return RetryConfig{
		MaxRetries:  3,
		BaseDelay:   100 * time.Millisecond,
		MaxDelay:    5 * time.Second,
		Multiplier:  2.0,
		JitterRatio: 0.1,
	}
}

// RetryWithBackoff executes fn with exponential backoff and jitter.
func RetryWithBackoff(ctx context.Context, cfg RetryConfig, fn func(context.Context) error) error {
	var lastErr error
	for attempt := 0; attempt <= cfg.MaxRetries; attempt++ {
		if err := ctx.Err(); err != nil {
			return fmt.Errorf("context cancelled during retry: %w", err)
		}

		lastErr = fn(ctx)
		if lastErr == nil {
			return nil
		}

		if attempt == cfg.MaxRetries {
			break
		}

		delay := float64(cfg.BaseDelay) * math.Pow(cfg.Multiplier, float64(attempt))
		if delay > float64(cfg.MaxDelay) {
			delay = float64(cfg.MaxDelay)
		}
		jitter := delay * cfg.JitterRatio * (rand.Float64()*2 - 1)
		sleepDuration := time.Duration(delay + jitter)

		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(sleepDuration):
		}
	}
	return fmt.Errorf("%w: %v", ErrTooManyRetries, lastErr)
}

// ResilientHTTPClient wraps http.Client with circuit breaker and retry logic.
type ResilientHTTPClient struct {
	client  *http.Client
	breaker *CircuitBreaker
	retry   RetryConfig
	logger  *zap.Logger
}

// NewResilientHTTPClient creates an HTTP client with circuit breaker and retries.
func NewResilientHTTPClient(logger *zap.Logger, serviceName string) *ResilientHTTPClient {
	return &ResilientHTTPClient{
		client: &http.Client{
			Timeout: 10 * time.Second,
			Transport: &http.Transport{
				MaxIdleConns:        100,
				MaxIdleConnsPerHost: 10,
				IdleConnTimeout:     90 * time.Second,
			},
		},
		breaker: NewCircuitBreaker(logger, DefaultCircuitBreakerConfig(serviceName)),
		retry:   DefaultRetryConfig(),
		logger:  logger,
	}
}

// Do executes an HTTP request with circuit breaker and retry logic.
func (c *ResilientHTTPClient) Do(ctx context.Context, req *http.Request) (*http.Response, error) {
	var resp *http.Response
	err := c.breaker.Execute(ctx, func(ctx context.Context) error {
		return RetryWithBackoff(ctx, c.retry, func(ctx context.Context) error {
			req = req.WithContext(ctx)
			var err error
			resp, err = c.client.Do(req)
			if err != nil {
				return err
			}
			if resp.StatusCode >= 500 {
				return fmt.Errorf("server error: %d", resp.StatusCode)
			}
			return nil
		})
	})
	return resp, err
}

// BreakerState returns the current state of the circuit breaker.
func (c *ResilientHTTPClient) BreakerState() string {
	return c.breaker.State()
}

File: infrastructure/rust-sdk/src/circuit_breaker.rs
//! Circuit breaker and retry with exponential backoff for Rust services.

use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::Mutex;

#[derive(Debug, Clone, PartialEq)]
pub enum CircuitState {
    Closed,
    Open,
    HalfOpen,
}

impl std::fmt::Display for CircuitState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CircuitState::Closed => write!(f, "closed"),
            CircuitState::Open => write!(f, "open"),
            CircuitState::HalfOpen => write!(f, "half-open"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    pub name: String,
    pub failure_threshold: u32,
    pub success_threshold: u32,
    pub timeout: Duration,
    pub half_open_max: u32,
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        Self {
            name: "default".to_string(),
            failure_threshold: 5,
            success_threshold: 3,
            timeout: Duration::from_secs(30),
            half_open_max: 1,
        }
    }
}

impl CircuitBreakerConfig {
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            ..Default::default()
        }
    }
}

struct CircuitBreakerInner {
    state: CircuitState,
    failure_count: u32,
    success_count: u32,
    last_failure: Option<Instant>,
    half_open_current: u32,
}

/// Implements the circuit breaker pattern for inter-service calls.
pub struct CircuitBreaker {
    config: CircuitBreakerConfig,
    inner: Arc<Mutex<CircuitBreakerInner>>,
}

impl CircuitBreaker {
    pub fn new(config: CircuitBreakerConfig) -> Self {
        Self {
            config,
            inner: Arc::new(Mutex::new(CircuitBreakerInner {
                state: CircuitState::Closed,
                failure_count: 0,
                success_count: 0,
                last_failure: None,
                half_open_current: 0,
            })),
        }
    }

    pub async fn execute<F, T, E>(&self, f: F) -> Result<T, CircuitBreakerError<E>>
    where
        F: std::future::Future<Output = Result<T, E>>,
    {
        {
            let mut inner = self.inner.lock().await;
            if !self.can_execute(&mut inner) {
                return Err(CircuitBreakerError::Open(self.config.name.clone()));
            }
        }

        match f.await {
            Ok(result) => {
                self.record_success().await;
                Ok(result)
            }
            Err(e) => {
                self.record_failure().await;
                Err(CircuitBreakerError::Inner(e))
            }
        }
    }

    pub async fn state(&self) -> CircuitState {
        self.inner.lock().await.state.clone()
    }

    fn can_execute(&self, inner: &mut CircuitBreakerInner) -> bool {
        match inner.state {
            CircuitState::Closed => true,
            CircuitState::Open => {
                if let Some(last) = inner.last_failure {
                    if last.elapsed() > self.config.timeout {
                        inner.state = CircuitState::HalfOpen;
                        inner.half_open_current = 0;
                        inner.success_count = 0;
                        return true;
                    }
                }
                false
            }
            CircuitState::HalfOpen => {
                if inner.half_open_current >= self.config.half_open_max {
                    return false;
                }
                inner.half_open_current += 1;
                true
            }
        }
    }

    async fn record_success(&self) {
        let mut inner = self.inner.lock().await;
        match inner.state {
            CircuitState::HalfOpen => {
                inner.success_count += 1;
                if inner.success_count >= self.config.success_threshold {
                    inner.state = CircuitState::Closed;
                    inner.failure_count = 0;
                    inner.success_count = 0;
                }
            }
            _ => {
                inner.failure_count = 0;
            }
        }
    }

    async fn record_failure(&self) {
        let mut inner = self.inner.lock().await;
        inner.failure_count += 1;
        inner.last_failure = Some(Instant::now());

        match inner.state {
            CircuitState::HalfOpen => {
                inner.state = CircuitState::Open;
            }
            CircuitState::Closed => {
                if inner.failure_count >= self.config.failure_threshold {
                    inner.state = CircuitState::Open;
                }
            }
            _ => {}
        }
    }
}

#[derive(Debug)]
pub enum CircuitBreakerError<E> {
    Open(String),
    Inner(E),
}

impl<E: std::fmt::Display> std::fmt::Display for CircuitBreakerError<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CircuitBreakerError::Open(name) => write!(f, "circuit breaker '{}' is open", name),
            CircuitBreakerError::Inner(e) => write!(f, "{}", e),
        }
    }
}

impl<E: std::fmt::Debug + std::fmt::Display> std::error::Error for CircuitBreakerError<E> {}

#[derive(Debug, Clone)]
pub struct RetryConfig {
    pub max_retries: u32,
    pub base_delay: Duration,
    pub max_delay: Duration,
    pub multiplier: f64,
    pub jitter_ratio: f64,
}

impl Default for RetryConfig {
    fn default() -> Self {
        Self {
            max_retries: 3,
            base_delay: Duration::from_millis(100),
            max_delay: Duration::from_secs(5),
            multiplier: 2.0,
            jitter_ratio: 0.1,
        }
    }
}

/// Execute an async function with exponential backoff and jitter.
pub async fn retry_with_backoff<F, Fut, T, E>(
    config: &RetryConfig,
    mut f: F,
) -> Result<T, E>
where
    F: FnMut() -> Fut,
    Fut: std::future::Future<Output = Result<T, E>>,
{
    let mut last_error: Option<E> = None;

    for attempt in 0..=config.max_retries {
        match f().await {
            Ok(result) => return Ok(result),
            Err(e) => {
                last_error = Some(e);
                if attempt == config.max_retries {
                    break;
                }

                let delay_ms = (config.base_delay.as_millis() as f64)
                    * config.multiplier.powi(attempt as i32);
                let delay_ms = delay_ms.min(config.max_delay.as_millis() as f64);
                let jitter = delay_ms * config.jitter_ratio * (rand_f64() * 2.0 - 1.0);
                let sleep_ms = (delay_ms + jitter).max(0.0) as u64;

                tokio::time::sleep(Duration::from_millis(sleep_ms)).await;
            }
        }
    }

    Err(last_error.expect("retry loop must record at least one error"))
}

fn rand_f64() -> f64 {
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};
    let mut hasher = DefaultHasher::new();
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos()
        .hash(&mut hasher);
    (hasher.finish() % 10000) as f64 / 10000.0
}
